From b7eef26595cd64d3b75706c473cdb2b77fd3a1e4 Mon Sep 17 00:00:00 2001
From: Chao Yu <chao@kernel.org>
Date: Wed, 5 Nov 2025 16:16:31 +0800
Subject: [PATCH 46/49] backport: f2fs: fix to do sanity check on node footer
 for non inode dnode

As syzbot reported below:

------------[ cut here ]------------
kernel BUG at fs/f2fs/file.c:1243!
Oops: invalid opcode: 0000 [#1] SMP KASAN NOPTI
CPU: 0 UID: 0 PID: 5354 Comm: syz.0.0 Not tainted 6.17.0-rc1-syzkaller-00211-g90d970cade8e #0 PREEMPT(full)
RIP: 0010:f2fs_truncate_hole+0x69e/0x6c0 fs/f2fs/file.c:1243
Call Trace:
 <TASK>
 f2fs_punch_hole+0x2db/0x330 fs/f2fs/file.c:1306
 f2fs_fallocate+0x546/0x990 fs/f2fs/file.c:2018
 vfs_fallocate+0x666/0x7e0 fs/open.c:342
 ksys_fallocate fs/open.c:366 [inline]
 __do_sys_fallocate fs/open.c:371 [inline]
 __se_sys_fallocate fs/open.c:369 [inline]
 __x64_sys_fallocate+0xc0/0x110 fs/open.c:369
 do_syscall_x64 arch/x86/entry/syscall_64.c:63 [inline]
 do_syscall_64+0xfa/0x3b0 arch/x86/entry/syscall_64.c:94
 entry_SYSCALL_64_after_hwframe+0x77/0x7f
RIP: 0033:0x7f1e65f8ebe9

w/ a fuzzed image, f2fs may encounter panic due to it detects inconsistent
truncation range in direct node in f2fs_truncate_hole().

The root cause is: a non-inode dnode may has the same footer.ino and
footer.nid, so the dnode will be parsed as an inode, then ADDRS_PER_PAGE()
may return wrong blkaddr count which may be 923 typically, by chance,
dn.ofs_in_node is equal to 923, then count can be calculated to 0 in below
statement, later it will trigger panic w/ f2fs_bug_on(, count == 0 || ...).

	count = min(end_offset - dn.ofs_in_node, pg_end - pg_start);

This patch introduces a new node_type NODE_TYPE_NON_INODE, then allowing
passing the new_type to sanity_check_node_footer in f2fs_get_node_folio()
to detect corruption that a non-inode dnode has the same footer.ino and
footer.nid.

Scripts to reproduce:
mkfs.f2fs -f /dev/vdb
mount /dev/vdb /mnt/f2fs
touch /mnt/f2fs/foo
touch /mnt/f2fs/bar
dd if=/dev/zero of=/mnt/f2fs/foo bs=1M count=8
umount /mnt/f2fs
inject.f2fs --node --mb i_nid --nid 4 --idx 0 --val 5 /dev/vdb
mount /dev/vdb /mnt/f2fs
xfs_io /mnt/f2fs/foo -c "fpunch 6984k 4k"

Cc: stable@kernel.org
Reported-by: syzbot+b9c7ffd609c3f09416ab@syzkaller.appspotmail.com
Closes: https://lore.kernel.org/linux-f2fs-devel/68a68e27.050a0220.1a3988.0002.GAE@google.com
Signed-off-by: Chao Yu <chao@kernel.org>
Signed-off-by: Jaegeuk Kim <jaegeuk@kernel.org>
---
 fs/f2fs/data.c     |  6 ++--
 fs/f2fs/dir.c      |  2 +-
 fs/f2fs/f2fs.h     |  4 ++-
 fs/f2fs/file.c     |  2 +-
 fs/f2fs/gc.c       |  4 +--
 fs/f2fs/inline.c   | 20 ++++++-------
 fs/f2fs/inode.c    |  4 +--
 fs/f2fs/node.c     | 75 +++++++++++++++++++++++++++++++++-------------
 fs/f2fs/node.h     |  7 +++++
 fs/f2fs/recovery.c |  2 +-
 fs/f2fs/xattr.c    |  9 +++---
 11 files changed, 90 insertions(+), 45 deletions(-)

diff --git a/fs/f2fs/data.c b/fs/f2fs/data.c
index 859db4dfba91..be4fef0f7245 100644
--- a/fs/f2fs/data.c
+++ b/fs/f2fs/data.c
@@ -3339,7 +3339,7 @@ static int prepare_write_begin(struct f2fs_sb_info *sbi,

 restart:
 	/* check inline_data */
-	ipage = f2fs_get_node_page(sbi, inode->i_ino);
+	ipage = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage)) {
 		err = PTR_ERR(ipage);
 		goto unlock_out;
@@ -3398,7 +3398,7 @@ static int __find_data_block(struct inode *inode, pgoff_t index,
 	struct extent_info ei = {0, };
 	int err = 0;

-	ipage = f2fs_get_node_page(F2FS_I_SB(inode), inode->i_ino);
+	ipage = f2fs_get_node_page(F2FS_I_SB(inode), inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage))
 		return PTR_ERR(ipage);

@@ -3429,7 +3429,7 @@ static int __reserve_data_block(struct inode *inode, pgoff_t index,

 	f2fs_do_map_lock(sbi, F2FS_GET_BLOCK_PRE_AIO, true);

-	ipage = f2fs_get_node_page(sbi, inode->i_ino);
+	ipage = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage)) {
 		err = PTR_ERR(ipage);
 		goto unlock_out;
diff --git a/fs/f2fs/dir.c b/fs/f2fs/dir.c
index 1f74508eaf6f..36bb67ad031a 100644
--- a/fs/f2fs/dir.c
+++ b/fs/f2fs/dir.c
@@ -639,7 +639,7 @@ struct page *f2fs_init_inode_metadata(struct inode *inode, struct inode *dir,
 				goto put_error;
 		}
 	} else {
-		page = f2fs_get_node_page(F2FS_I_SB(dir), inode->i_ino);
+		page = f2fs_get_node_page(F2FS_I_SB(dir), inode->i_ino, NODE_TYPE_REGULAR);
 		if (IS_ERR(page))
 			return page;
 	}
diff --git a/fs/f2fs/f2fs.h b/fs/f2fs/f2fs.h
index 869bc8fbbcc8..9d4915b387c1 100644
--- a/fs/f2fs/f2fs.h
+++ b/fs/f2fs/f2fs.h
@@ -3584,6 +3584,7 @@ void f2fs_hash_filename(const struct inode *dir, struct f2fs_filename *fname);
  * node.c
  */
 struct node_info;
+enum node_type;

 int f2fs_check_nid_range(struct f2fs_sb_info *sbi, nid_t nid);
 bool f2fs_available_free_memory(struct f2fs_sb_info *sbi, int type);
@@ -3607,7 +3608,8 @@ int f2fs_remove_inode_page(struct inode *inode);
 struct page *f2fs_new_inode_page(struct inode *inode);
 struct page *f2fs_new_node_page(struct dnode_of_data *dn, unsigned int ofs);
 void f2fs_ra_node_page(struct f2fs_sb_info *sbi, nid_t nid);
-struct page *f2fs_get_node_page(struct f2fs_sb_info *sbi, pgoff_t nid);
+struct page *f2fs_get_node_page(struct f2fs_sb_info *sbi, pgoff_t nid,
+					enum node_type node_type);
 struct page *f2fs_get_node_page_ra(struct page *parent, int start);
 int f2fs_move_node_page(struct page *node_page, int gc_type);
 void f2fs_flush_inline_data(struct f2fs_sb_info *sbi);
diff --git a/fs/f2fs/file.c b/fs/f2fs/file.c
index 1c711c829b22..aff5c452a4d8 100644
--- a/fs/f2fs/file.c
+++ b/fs/f2fs/file.c
@@ -705,7 +705,7 @@ int f2fs_do_truncate_blocks(struct inode *inode, u64 from, bool lock)
 	if (lock)
 		f2fs_lock_op(sbi);

-	ipage = f2fs_get_node_page(sbi, inode->i_ino);
+	ipage = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage)) {
 		err = PTR_ERR(ipage);
 		goto out;
diff --git a/fs/f2fs/gc.c b/fs/f2fs/gc.c
index 1208f7ec8588..309d042dcc57 100644
--- a/fs/f2fs/gc.c
+++ b/fs/f2fs/gc.c
@@ -999,7 +999,7 @@ static int gc_node_segment(struct f2fs_sb_info *sbi,
 		}

 		/* phase == 2 */
-		node_page = f2fs_get_node_page(sbi, nid);
+		node_page = f2fs_get_node_page(sbi, nid, NODE_TYPE_REGULAR);
 		if (IS_ERR(node_page))
 			continue;

@@ -1073,7 +1073,7 @@ static bool is_alive(struct f2fs_sb_info *sbi, struct f2fs_summary *sum,
 	nid = le32_to_cpu(sum->nid);
 	ofs_in_node = le16_to_cpu(sum->ofs_in_node);

-	node_page = f2fs_get_node_page(sbi, nid);
+	node_page = f2fs_get_node_page(sbi, nid, NODE_TYPE_REGULAR);
 	if (IS_ERR(node_page))
 		return false;

diff --git a/fs/f2fs/inline.c b/fs/f2fs/inline.c
index bdc7d0339f96..957efed2e6a8 100644
--- a/fs/f2fs/inline.c
+++ b/fs/f2fs/inline.c
@@ -113,7 +113,7 @@ int f2fs_read_inline_data(struct inode *inode, struct page *page)
 						path, current->comm);
 	}

-	ipage = f2fs_get_node_page(F2FS_I_SB(inode), inode->i_ino);
+	ipage = f2fs_get_node_page(F2FS_I_SB(inode), inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage)) {
 		trace_android_fs_dataread_end(inode, page_offset(page),
 					      PAGE_SIZE);
@@ -238,7 +238,7 @@ int f2fs_convert_inline_inode(struct inode *inode)

 	f2fs_lock_op(sbi);

-	ipage = f2fs_get_node_page(sbi, inode->i_ino);
+	ipage = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage)) {
 		err = PTR_ERR(ipage);
 		goto out;
@@ -314,7 +314,7 @@ int f2fs_recover_inline_data(struct inode *inode, struct page *npage)
 	if (f2fs_has_inline_data(inode) &&
 			ri && (ri->i_inline & F2FS_INLINE_DATA)) {
 process_inline:
-		ipage = f2fs_get_node_page(sbi, inode->i_ino);
+		ipage = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 		if (IS_ERR(ipage))
 			return PTR_ERR(ipage);

@@ -333,7 +333,7 @@ int f2fs_recover_inline_data(struct inode *inode, struct page *npage)
 	}

 	if (f2fs_has_inline_data(inode)) {
-		ipage = f2fs_get_node_page(sbi, inode->i_ino);
+		ipage = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 		if (IS_ERR(ipage))
 			return PTR_ERR(ipage);
 		f2fs_truncate_inline_inode(inode, ipage, 0);
@@ -363,7 +363,7 @@ struct f2fs_dir_entry *f2fs_find_in_inline_dir(struct inode *dir,
 	struct page *ipage;
 	void *inline_dentry;

-	ipage = f2fs_get_node_page(sbi, dir->i_ino);
+	ipage = f2fs_get_node_page(sbi, dir->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage)) {
 		*res_page = ipage;
 		return NULL;
@@ -611,7 +611,7 @@ int f2fs_try_convert_inline_dir(struct inode *dir, struct dentry *dentry)
 	if (err)
 		goto out;

-	ipage = f2fs_get_node_page(sbi, dir->i_ino);
+	ipage = f2fs_get_node_page(sbi, dir->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage)) {
 		err = PTR_ERR(ipage);
 		goto out_fname;
@@ -646,7 +646,7 @@ int f2fs_add_inline_entry(struct inode *dir, const struct f2fs_filename *fname,
 	struct page *page = NULL;
 	int err = 0;

-	ipage = f2fs_get_node_page(sbi, dir->i_ino);
+	ipage = f2fs_get_node_page(sbi, dir->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage))
 		return PTR_ERR(ipage);

@@ -736,7 +736,7 @@ bool f2fs_empty_inline_dir(struct inode *dir)
 	void *inline_dentry;
 	struct f2fs_dentry_ptr d;

-	ipage = f2fs_get_node_page(sbi, dir->i_ino);
+	ipage = f2fs_get_node_page(sbi, dir->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage))
 		return false;

@@ -767,7 +767,7 @@ int f2fs_read_inline_dir(struct file *file, struct dir_context *ctx,
 	if (ctx->pos == d.max)
 		return 0;

-	ipage = f2fs_get_node_page(F2FS_I_SB(inode), inode->i_ino);
+	ipage = f2fs_get_node_page(F2FS_I_SB(inode), inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage))
 		return PTR_ERR(ipage);

@@ -799,7 +799,7 @@ int f2fs_inline_data_fiemap(struct inode *inode,
 	struct page *ipage;
 	int err = 0;

-	ipage = f2fs_get_node_page(F2FS_I_SB(inode), inode->i_ino);
+	ipage = f2fs_get_node_page(F2FS_I_SB(inode), inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage))
 		return PTR_ERR(ipage);

diff --git a/fs/f2fs/inode.c b/fs/f2fs/inode.c
index 449ee4fd3ac1..86365ba8a1c6 100644
--- a/fs/f2fs/inode.c
+++ b/fs/f2fs/inode.c
@@ -367,7 +367,7 @@ static int do_read_inode(struct inode *inode)
 	if (f2fs_check_nid_range(sbi, inode->i_ino))
 		return -EINVAL;

-	node_page = f2fs_get_node_page(sbi, inode->i_ino);
+	node_page = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(node_page))
 		return PTR_ERR(node_page);

@@ -728,7 +728,7 @@ void f2fs_update_inode_page(struct inode *inode)
 	struct f2fs_sb_info *sbi = F2FS_I_SB(inode);
 	struct page *node_page;
 retry:
-	node_page = f2fs_get_node_page(sbi, inode->i_ino);
+	node_page = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(node_page)) {
 		int err = PTR_ERR(node_page);

diff --git a/fs/f2fs/node.c b/fs/f2fs/node.c
index 7825cb6d1587..7bbb84236abd 100644
--- a/fs/f2fs/node.c
+++ b/fs/f2fs/node.c
@@ -773,7 +773,7 @@ int f2fs_get_dnode_of_data(struct dnode_of_data *dn, pgoff_t index, int mode)
 	npage[0] = dn->inode_page;

 	if (!npage[0]) {
-		npage[0] = f2fs_get_node_page(sbi, nids[0]);
+		npage[0] = f2fs_get_node_page(sbi, nids[0], NODE_TYPE_REGULAR);
 		if (IS_ERR(npage[0]))
 			return PTR_ERR(npage[0]);
 	}
@@ -839,7 +839,7 @@ int f2fs_get_dnode_of_data(struct dnode_of_data *dn, pgoff_t index, int mode)
 		}

 		if (!done) {
-			npage[i] = f2fs_get_node_page(sbi, nids[i]);
+			npage[i] = f2fs_get_node_page(sbi, nids[i], NODE_TYPE_NON_INODE);
 			if (IS_ERR(npage[i])) {
 				err = PTR_ERR(npage[i]);
 				f2fs_put_page(npage[0], 0);
@@ -938,7 +938,7 @@ static int truncate_dnode(struct dnode_of_data *dn)
 		return 1;

 	/* get direct node */
-	page = f2fs_get_node_page(F2FS_I_SB(dn->inode), dn->nid);
+	page = f2fs_get_node_page(F2FS_I_SB(dn->inode), dn->nid, NODE_TYPE_NON_INODE);
 	if (PTR_ERR(page) == -ENOENT)
 		return 1;
 	else if (IS_ERR(page))
@@ -973,7 +973,7 @@ static int truncate_nodes(struct dnode_of_data *dn, unsigned int nofs,

 	trace_f2fs_truncate_nodes_enter(dn->inode, dn->nid, dn->data_blkaddr);

-	page = f2fs_get_node_page(F2FS_I_SB(dn->inode), dn->nid);
+	page = f2fs_get_node_page(F2FS_I_SB(dn->inode), dn->nid, NODE_TYPE_NON_INODE);
 	if (IS_ERR(page)) {
 		trace_f2fs_truncate_nodes_exit(dn->inode, PTR_ERR(page));
 		return PTR_ERR(page);
@@ -1051,7 +1051,7 @@ static int truncate_partial_nodes(struct dnode_of_data *dn,
 	/* get indirect nodes in the path */
 	for (i = 0; i < idx + 1; i++) {
 		/* reference count'll be increased */
-		pages[i] = f2fs_get_node_page(F2FS_I_SB(dn->inode), nid[i]);
+		pages[i] = f2fs_get_node_page(F2FS_I_SB(dn->inode), nid[i], NODE_TYPE_NON_INODE);
 		if (IS_ERR(pages[i])) {
 			err = PTR_ERR(pages[i]);
 			idx = i - 1;
@@ -1124,7 +1124,7 @@ int f2fs_truncate_inode_blocks(struct inode *inode, pgoff_t from)
 		return level;
 	}

-	page = f2fs_get_node_page(sbi, inode->i_ino);
+	page = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(page)) {
 		trace_f2fs_truncate_inode_blocks_exit(inode, PTR_ERR(page));
 		return PTR_ERR(page);
@@ -1215,7 +1215,7 @@ int f2fs_truncate_xattr_node(struct inode *inode)
 	if (!nid)
 		return 0;

-	npage = f2fs_get_node_page(sbi, nid);
+	npage = f2fs_get_node_page(sbi, nid, NODE_TYPE_REGULAR);
 	if (IS_ERR(npage))
 		return PTR_ERR(npage);

@@ -1419,8 +1419,46 @@ void f2fs_ra_node_page(struct f2fs_sb_info *sbi, nid_t nid)
 	f2fs_put_page(apage, err ? 1 : 0);
 }

+static int sanity_check_node_footer(struct f2fs_sb_info *sbi,
+				struct page *page, pgoff_t nid,
+				enum node_type ntype)
+{
+	if (unlikely(nid != nid_of_node(page)))
+		goto out_err;
+
+	switch (ntype) {
+	case NODE_TYPE_INODE:
+		if (!IS_INODE(page))
+			goto out_err;
+		break;
+	case NODE_TYPE_XATTR:
+		if (!f2fs_has_xattr_block(ofs_of_node(page)))
+			goto out_err;
+		break;
+	case NODE_TYPE_NON_INODE:
+		if (IS_INODE(page))
+			goto out_err;
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+out_err:
+	f2fs_warn(sbi, "inconsistent node block, node_type:%d, nid:%lu, "
+		  "node_footer[nid:%u,ino:%u,ofs:%u,cpver:%llu,blkaddr:%u]",
+		  ntype, nid, nid_of_node(page), ino_of_node(page),
+		  ofs_of_node(page), cpver_of_node(page),
+		  next_blkaddr_of_node(page));
+	set_sbi_flag(sbi, SBI_NEED_FSCK);
+	f2fs_handle_error(sbi, ERROR_INCONSISTENT_FOOTER);
+	return -EFSCORRUPTED;
+}
+
+
+
 static struct page *__get_node_page(struct f2fs_sb_info *sbi, pgoff_t nid,
-					struct page *parent, int start)
+					struct page *parent, int start, enum node_type node_type)
 {
 	struct page *page;
 	int err;
@@ -1462,15 +1500,11 @@ static struct page *__get_node_page(struct f2fs_sb_info *sbi, pgoff_t nid,
 		goto out_err;
 	}
 page_hit:
-	if (likely(nid == nid_of_node(page)))
-		return page;
+	err = sanity_check_node_footer(sbi, page, nid, node_type);
+	if (err)
+		goto out_err;
+	return page;

-	f2fs_warn(sbi, "inconsistent node block, nid:%lu, node_footer[nid:%u,ino:%u,ofs:%u,cpver:%llu,blkaddr:%u]",
-			  nid, nid_of_node(page), ino_of_node(page),
-			  ofs_of_node(page), cpver_of_node(page),
-			  next_blkaddr_of_node(page));
-	set_sbi_flag(sbi, SBI_NEED_FSCK);
-	err = -EINVAL;
 out_err:
 	ClearPageUptodate(page);
 out_put_err:
@@ -1481,9 +1515,10 @@ static struct page *__get_node_page(struct f2fs_sb_info *sbi, pgoff_t nid,
 	return ERR_PTR(err);
 }

-struct page *f2fs_get_node_page(struct f2fs_sb_info *sbi, pgoff_t nid)
+struct page *f2fs_get_node_page(struct f2fs_sb_info *sbi, pgoff_t nid,
+					enum node_type node_type)
 {
-	return __get_node_page(sbi, nid, NULL, 0);
+	return __get_node_page(sbi, nid, NULL, 0, node_type);
 }

 struct page *f2fs_get_node_page_ra(struct page *parent, int start)
@@ -1491,7 +1526,7 @@ struct page *f2fs_get_node_page_ra(struct page *parent, int start)
 	struct f2fs_sb_info *sbi = F2FS_P_SB(parent);
 	nid_t nid = get_nid(parent, start, false);

-	return __get_node_page(sbi, nid, parent, start);
+	return __get_node_page(sbi, nid, parent, start, NODE_TYPE_REGULAR);
 }

 static void flush_inline_data(struct f2fs_sb_info *sbi, nid_t ino)
@@ -2686,7 +2721,7 @@ int f2fs_recover_inline_xattr(struct inode *inode, struct page *page)
 	struct page *ipage;
 	struct f2fs_inode *ri;

-	ipage = f2fs_get_node_page(F2FS_I_SB(inode), inode->i_ino);
+	ipage = f2fs_get_node_page(F2FS_I_SB(inode), inode->i_ino, NODE_TYPE_REGULAR);
 	if (IS_ERR(ipage))
 		return PTR_ERR(ipage);

diff --git a/fs/f2fs/node.h b/fs/f2fs/node.h
index 3c09cae058b0..aadd220b59ad 100644
--- a/fs/f2fs/node.h
+++ b/fs/f2fs/node.h
@@ -53,6 +53,13 @@ enum {
 	IS_PREALLOC,		/* nat entry is preallocated */
 };

+enum node_type {
+	NODE_TYPE_REGULAR,
+	NODE_TYPE_INODE,
+	NODE_TYPE_XATTR,
+	NODE_TYPE_NON_INODE,
+};
+
 /*
  * For node information
  */
diff --git a/fs/f2fs/recovery.c b/fs/f2fs/recovery.c
index 0dd0db553487..06c3f9f12a56 100644
--- a/fs/f2fs/recovery.c
+++ b/fs/f2fs/recovery.c
@@ -523,7 +523,7 @@ static int check_index_in_prev_nodes(struct f2fs_sb_info *sbi,
 	}

 	/* Get the node page */
-	node_page = f2fs_get_node_page(sbi, nid);
+	node_page = f2fs_get_node_page(sbi, nid, NODE_TYPE_REGULAR);
 	if (IS_ERR(node_page))
 		return PTR_ERR(node_page);

diff --git a/fs/f2fs/xattr.c b/fs/f2fs/xattr.c
index 0005d273da44..96856de846e1 100644
--- a/fs/f2fs/xattr.c
+++ b/fs/f2fs/xattr.c
@@ -22,6 +22,7 @@
 #include "f2fs.h"
 #include "xattr.h"
 #include "segment.h"
+#include "node.h"

 static void *xattr_alloc(struct f2fs_sb_info *sbi, int size, bool *is_inline)
 {
@@ -281,7 +282,7 @@ static int read_inline_xattr(struct inode *inode, struct page *ipage,
 	if (ipage) {
 		inline_addr = inline_xattr_addr(inode, ipage);
 	} else {
-		page = f2fs_get_node_page(sbi, inode->i_ino);
+		page = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 		if (IS_ERR(page))
 			return PTR_ERR(page);

@@ -302,7 +303,7 @@ static int read_xattr_block(struct inode *inode, void *txattr_addr)
 	void *xattr_addr;

 	/* The inode already has an extended attribute block. */
-	xpage = f2fs_get_node_page(sbi, xnid);
+	xpage = f2fs_get_node_page(sbi, xnid, NODE_TYPE_REGULAR);
 	if (IS_ERR(xpage))
 		return PTR_ERR(xpage);

@@ -448,7 +449,7 @@ static inline int write_all_xattrs(struct inode *inode, __u32 hsize,
 		if (ipage) {
 			inline_addr = inline_xattr_addr(inode, ipage);
 		} else {
-			in_page = f2fs_get_node_page(sbi, inode->i_ino);
+			in_page = f2fs_get_node_page(sbi, inode->i_ino, NODE_TYPE_REGULAR);
 			if (IS_ERR(in_page)) {
 				f2fs_alloc_nid_failed(sbi, new_nid);
 				return PTR_ERR(in_page);
@@ -474,7 +475,7 @@ static inline int write_all_xattrs(struct inode *inode, __u32 hsize,

 	/* write to xattr node block */
 	if (F2FS_I(inode)->i_xattr_nid) {
-		xpage = f2fs_get_node_page(sbi, F2FS_I(inode)->i_xattr_nid);
+		            xpage = f2fs_get_node_page(sbi, F2FS_I(inode)->i_xattr_nid, NODE_TYPE_REGULAR);
 		if (IS_ERR(xpage)) {
 			err = PTR_ERR(xpage);
 			f2fs_alloc_nid_failed(sbi, new_nid);
--
2.43.0

